core/lib/zksync_core/src/eth_watch/event_processors/governance_upgrades.rs

```diff
@@ -0,0 +1,117 @@
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl<W: EthClient + Sync> EventProcessor<W> for GovernanceUpgradesEventProcessor {
+    async fn process_events(
+        &mut self,
+        storage: &mut StorageProcessor<'_>,
+        client: &W,
+        events: Vec<Log>,
+    ) -> Result<(), Error> {
+        let mut upgrades = Vec::new();
+        for event in events
+            .into_iter()
+            .filter(|event| event.topics[0] == self.upgrade_proposal_signature)
+        {
+            let governance_operation = GovernanceOperation::try_from(event)
+                .map_err(|err| Error::LogParse(format!("{:?}", err)))?;
+            // Some calls can target other contracts than Diamond proxy, skip them.
+            for call in governance_operation
+                .calls
+                .into_iter()
+                .filter(|call| call.target == self.diamond_proxy_address)
+            {
+                let upgrade = ProtocolUpgrade::try_from(call)
+                    .map_err(|err| Error::LogParse(format!("{:?}", err)))?;
+                // Scheduler VK is not present in proposal event. It is hardcoded in verifier contract.
+                let scheduler_vk_hash = if let Some(address) = upgrade.verifier_address {
+                    Some(client.scheduler_vk_hash(address).await?)
+                } else {
+                    None
+                };
+                upgrades.push((upgrade, scheduler_vk_hash));
+            }
+        }
+
+        if upgrades.is_empty() {
+            return Ok(());
+        }
+
+        let ids_str: Vec<_> = upgrades
+            .iter()
+            .map(|(u, _)| format!("{}", u.id as u16))
+            .collect();
+        tracing::debug!("Received upgrades with ids: {}", ids_str.join(", "));
+
+        let new_upgrades: Vec<_> = upgrades
+            .into_iter()
+            .skip_while(|(v, _)| v.id as u16 <= self.last_seen_version_id as u16)
+            .collect();
+        if new_upgrades.is_empty() {
+            return Ok(());
+        }
+
+        let last_id = new_upgrades.last().unwrap().0.id;
+        let stage_start = Instant::now();
+        for (upgrade, scheduler_vk_hash) in new_upgrades {
+            let previous_version = storage
+                .protocol_versions_dal()
+                .load_previous_version(upgrade.id)
+                .await
+                .expect("Expected previous version to be present in DB");
+            let new_version = previous_version.apply_upgrade(upgrade, scheduler_vk_hash);
+            storage
+                .protocol_versions_dal()
+                .save_protocol_version_with_tx(new_version)
+                .await;
+        }
+        metrics::histogram!("eth_watcher.poll_eth_node", stage_start.elapsed(), "stage" => "persist_upgrades");
+
+        self.last_seen_version_id = last_id;
+
+        Ok(())
+    }
+
+    fn relevant_topic(&self) -> H256 {
+        self.upgrade_proposal_signature
+    }
+}
```

Comment:
ignorant question: Why wasn't this in main? Why is this boojum-specific?

Reply:
@vladbochok reworked how upgrades will be handled after boojum upgrade, there will be a new L1 contract called `Governor`. Do not confuse with `GovernanceFacet` that we have now -- it will be renamed to `AdminFacet`.

Reply:
@vladbochok do we have this documented somewhere?

Reply:
Resolving this to be able to merge, please ask @vladbochok directly